pkg/armrpc/asyncoperation/controller/result.go

```diff
@@ -40,20 +44,24 @@ func NewCanceledResult(message string) Result {
 }
 
 // NewFailedResult creates a new Result object with the given error details and sets the failed flag to true.
-func NewFailedResult(err v1.ErrorDetails) Result {
+func NewFailedResult(ctx context.Context, err error, errorDetails v1.ErrorDetails) Result {
 	r := Result{}
-	r.SetFailed(err, false)
+	r.SetFailed(ctx, err, errorDetails, false)
+
 	return r
 }
 
 // SetFailed sets the error response with Failed status.
-func (r *Result) SetFailed(err v1.ErrorDetails, requeue bool) {
+func (r *Result) SetFailed(ctx context.Context, err error, errorDetails v1.ErrorDetails, requeue bool) {
+	logger := ucplog.FromContextOrDiscard(ctx)
+	logger.Error(err, fmt.Sprintf("Operation Failed: %s", errorDetails.Message))
```

Comment:
As I mentioned above, if we want to log out for the failure, cancellation is also failure. we need to put the same code there. We might not want to have the same code here and there. So we have logged the failure or status in main worker loop. I think we first need to figure it out why the current logging is not working as expected.

Reply:
> I think we first need to figure it out why the current logging is not working as expected.

Some details on why it is not working is here: https://github.com/radius-project/radius/issues/7297#issuecomment-1989167067. Current design expects each controller to individually log an error, which isn't reliable as it requires specific knowledge and clearly hasn't worked well with misses in multiple places. The approach in this PR is one way to fix it. How would you recommend fixing this gap? If the root cause is something else according to your understanding, then please add details on that.

Reply:
> > I think we first need to figure it out why the current logging is not working as expected.
> 
> Some details on why it is not working is here: [#7297 (comment)](https://github.com/radius-project/radius/issues/7297#issuecomment-1989167067). Current design expects each controller to individually log an error, which isn't reliable as it requires specific knowledge and clearly hasn't worked well with misses in multiple places.

To be clear, we have logged error individually in the main worker loop, not in controller. When controller creates Result, it needs to be expected to return Result and error. then the following code must log out the error. This the centralized logging part for all controllers.

https://github.com/radius-project/radius/blob/c77088f794f18e7478b10b34ee3041f600b613d9/pkg/armrpc/asyncoperation/worker/worker.go#L253-L260

> The approach in this PR is one way to fix it. How would you recommend fixing this gap? If the root cause is something else according to your understanding, then please add details on that.

I think the PR is the workaround. We first need to understand the root cause and then have the right fix here. in this comments - https://github.com/radius-project/radius/issues/7297#issuecomment-1989167067, we might not be able to repro the same issue.

I do not know what happens. but it might have numerous reasons. For example, RP could crash before logging. Another hypothesis is that when workflow step collects the logs, the log might be missing. Functional tests save the RP logs by using kubectl logs, which is not reliable.

Reply:
@youngbupark Let me ask this to build a better understanding of the current state -

1. Can you show where the code is that logs the error caught [here](https://github.com/radius-project/radius/blob/main/pkg/portableresources/backend/controller/deleteresource.go#L98)?

2. Is this log statement unnecessary https://github.com/radius-project/radius/blob/main/pkg/portableresources/backend/controller/createorupdateresource.go#L100?

Reply:
> @youngbupark Let me ask this to build a better understanding of the current state -
> 
> 1. Can you share link to where the code is that logs the error caught [here](https://github.com/radius-project/radius/blob/main/pkg/portableresources/backend/controller/deleteresource.go#L98)?

This is a backend controller. asyncCtrl.Run calls `func (c *DeleteResource[P, T]) Run(ctx context.Context, request *ctrl.Request) (ctrl.Result, error) {`.

https://github.com/radius-project/radius/blob/2a681be2a36280a81177e5a1221f01accd33c89d/pkg/armrpc/asyncoperation/worker/worker.go#L253

For additional code flow, I recommend to debug the code with the debugger.

> 2. Is this log statement unnecessary https://github.com/radius-project/radius/blob/main/pkg/portableresources/backend/controller/createorupdateresource.go#L100?

This is a server-side code. It may or may not. This is the error logging in controller. If controller author thinks it is useful with the better message and print out the error earlier, they can add it.

Reply:
> > @youngbupark Let me ask this to build a better understanding of the current state -
> > 
> > 1. Can you share link to where the code is that logs the error caught [here](https://github.com/radius-project/radius/blob/main/pkg/portableresources/backend/controller/deleteresource.go#L98)?
> 
> This is a backend controller. asyncCtrl.Run calls `func (c *DeleteResource[P, T]) Run(ctx context.Context, request *ctrl.Request) (ctrl.Result, error) {`.
> 
> https://github.com/radius-project/radius/blob/2a681be2a36280a81177e5a1221f01accd33c89d/pkg/armrpc/asyncoperation/worker/worker.go#L253
> 
> For additional code flow, I recommend to debug the code with the debugger.

I'm aware of where it is called from. I'm asking where is the error logged for the scenario I linked above.

> > 2. Is this log statement unnecessary https://github.com/radius-project/radius/blob/main/pkg/portableresources/backend/controller/createorupdateresource.go#L100?
> 
> This is a server-side code. It may or may not. This is the error logging in controller. If controller author thinks it is useful with the better message and print out the error earlier, they can add it.

My question is - is this redundant?

Reply:
> > > @youngbupark Let me ask this to build a better understanding of the current state -
> > > 
> > > 1. Can you share link to where the code is that logs the error caught [here](https://github.com/radius-project/radius/blob/main/pkg/portableresources/backend/controller/deleteresource.go#L98)?
> > 
> > 
> > This is a backend controller. asyncCtrl.Run calls `func (c *DeleteResource[P, T]) Run(ctx context.Context, request *ctrl.Request) (ctrl.Result, error) {`.
> > https://github.com/radius-project/radius/blob/2a681be2a36280a81177e5a1221f01accd33c89d/pkg/armrpc/asyncoperation/worker/worker.go#L253
> > 
> > For additional code flow, I recommend to debug the code with the debugger.
> 
> I'm aware of where it is called from. I'm asking where is the error logged for the scenario I linked above.

There are two places:

https://github.com/radius-project/radius/blob/2a681be2a36280a81177e5a1221f01accd33c89d/pkg/armrpc/asyncoperation/worker/worker.go#L260

https://github.com/radius-project/radius/blob/2a681be2a36280a81177e5a1221f01accd33c89d/pkg/armrpc/asyncoperation/worker/worker.go#L270

And I strongly recommend to debug the code flow to understand the details.

> 
> > > 2. Is this log statement unnecessary https://github.com/radius-project/radius/blob/main/pkg/portableresources/backend/controller/createorupdateresource.go#L100?
> > 
> > 
> > This is a server-side code. It may or may not. This is the error logging in controller. If controller author thinks it is useful with the better message and print out the error earlier, they can add it.
> 
> My question is - is this redundant?

As I said, "It may or may not". In general, If the log is useful for debugging the server, then it should not be redundant. From what I read, it is more readable than just printing out pure go error. I would reach out to yetkin who contribute to this code.

Reply:
If we are logging the same error twice, then that is redundant imo. It was [added originally](https://github.com/radius-project/radius/pull/6276) because the error wasn't getting logged anywhere else (same problem that this PR was trying to solve).